contracts/combat/B5_Combat.cairo

```diff
@@ -0,0 +1,393 @@
+%lang starknet
+
+from starkware.cairo.common.alloc import alloc
+from starkware.cairo.common.bitwise import bitwise_and
+from starkware.cairo.common.cairo_builtins import BitwiseBuiltin
+from starkware.cairo.common.math import assert_not_zero, assert_le, split_int, unsigned_div_rem
+from starkware.cairo.common.pow import pow
+
+# a min delay between attacks on a Realm; it can't
+# be attacked again during cooldown
+const ATTACK_COOLDOWN_PERIOD = 86400  # 1 day
+
+const COMBAT_TYPE_ATTACK_VS_DEFENSE = 1
+const COMBAT_TYPE_WISDOM_VS_AGILITY = 2
+const COMBAT_OUTCOME_ATTACKER_WINS = 1
+const COMBAT_OUTCOME_DEFENDER_WINS = 2
+
+# namespace constants deliberately start at 1 to
+# 1) translate in a straighforward way to "human, 1-based index" land
+# 2) to differentiate between uninitialized value (i.e. 0 in Cairo)
+
+namespace TroopId:
+    const Watchman = 1
+    const Guard = 2
+    const GuardCaptain = 3
+    const Squire = 4
+    const Knight = 5
+    const KnightCommander = 6
+    const Scout = 7
+    const Archer = 8
+    const Sniper = 9
+    const Scorpio = 10
+    const Baillista = 11
+    const Catapult = 12
+    const Apprentice = 13
+    const Mage = 14
+    const Arcanist = 15
+    const GrandMarshal = 16
+end
+
+namespace TroopType:
+    const Melee = 1
+    const Ranged = 2
+    const Siege = 3
+end
+
+struct Troop:
+    member type : felt  # TroopType
+    member tier : felt
+    member agility : felt
+    member attack : felt
+    member defense : felt
+    member vitality : felt
+    member wisdom : felt
+end
+
+# TODO: maybe use vector from 0xnoncents to build the full squad?
```

Comment:
how do you see this working?

Reply:
Was just thinking if it could help in making the squad building more generic. In the end, I used normal Cairo arrays.